File: 2022/august/02-08-2022/kth-smallest-element-in-a-sorted-matrix.py
from heapq import heappush, heappop
from collections import defaultdict

class Solution:
    def kthSmallest(self, matrix, k: int) -> int:
        index = defaultdict(int)
        heap = []
        n = len(matrix)
        for i in range(len(matrix)):
            heappush(heap, (matrix[i][index[i]],i))
            index[i] += 1
        while heap and k > 0:
            cur, i = heappop(heap)
            if index[i] < n:
                heappush(heap, (matrix[i][index[i]],i))
            index[i] += 1
            k -= 1
        return cur


print(Solution().kthSmallest([[1,5,9],[10,11,13],[12,13,15]], 8))

File: 2022/august/02-08-2022/kth-smallest-element-in-a-sorted-matrix.rs
use std::collections::BinaryHeap;
impl Solution {
	pub fn kth_smallest(matrix: Vec<Vec<i32>>, k: i32) -> i32 {
		let mut pq = BinaryHeap::<i32>::new();
		for v in matrix.iter(){
			for x in v.iter() {pq.push(-x.clone());}
		}
		for i in (0..k-1) {pq.pop();}
		-pq.peek().cloned().unwrap()
	}
}
